src/java.security.jgss/macosx/native/libosxkrb5/SCDynamicStoreConfig.m

```diff
@@ -24,200 +24,147 @@
-
-        id value = [nsDict objectForKey:key];
-        jobject jvalue = [coercer coerceNSObject:value withEnv:env usingCoercer:coercer];
-
-        JNFCallObjectMethod(env, jHashTable, jm_Hashtable_put, jkey, jvalue);
-
-        if (jkey != NULL) (*env)->DeleteLocalRef(env, jkey);
-        if (jvalue != NULL) (*env)->DeleteLocalRef(env, jvalue);
-    }
-
-    return jHashTable;
-}
-
-- (id) coerceJavaObject:(jobject)obj withEnv:(JNIEnv *)env usingCoercer:(JNFTypeCoercion *)coercer {
-    return nil;
-}
-
-@end
-
-
-
-NSDictionary *realmConfigsForRealms(SCDynamicStoreRef store, NSArray *realms) {
-    NSMutableDictionary *dict = [NSMutableDictionary dictionary];
-
-    for (NSString *realm in realms) {
-        CFTypeRef realmInfo = SCDynamicStoreCopyValue(store, (CFStringRef) [NSString stringWithFormat:@"Kerberos:%@", realm]);
-
-        if (realmInfo == NULL || CFGetTypeID(realmInfo) != CFDictionaryGetTypeID()) {
-            if (realmInfo) CFRelease(realmInfo);
-            return nil;
-        }
-
-        [dict setObject:(NSArray *)realmInfo forKey:realm];
-        CFRelease(realmInfo);
-    }
-
-    return dict;
-}
-
+#include "jni.h"
+#include "jni_util.h"
 
 #define KERBEROS_DEFAULT_REALMS @"Kerberos-Default-Realms"
 #define KERBEROS_DEFAULT_REALM_MAPPINGS @"Kerberos-Domain-Realm-Mappings"
+#define KERBEROS_REALM_INFO @"Kerberos:%@"
+
+JavaVM *localVM;
 
 void _SCDynamicStoreCallBack(SCDynamicStoreRef store, CFArrayRef changedKeys, void *info) {
-   NSArray *keys = (NSArray *)changedKeys;
+    NSArray *keys = (NSArray *)changedKeys;
     if ([keys count] == 0) return;
     if (![keys containsObject:KERBEROS_DEFAULT_REALMS] && ![keys containsObject:KERBEROS_DEFAULT_REALM_MAPPINGS]) return;
-
-    JNFPerformEnvBlock(JNFThreadDetachOnThreadDeath | JNFThreadSetSystemClassLoaderOnAttach | JNFThreadAttachAsDaemon, ^(JNIEnv *env) {
-        static JNF_CLASS_CACHE(jc_Config, "sun/security/krb5/Config");
-        static JNF_STATIC_MEMBER_CACHE(jm_Config_refresh, jc_Config, "refresh", "()V");
-        JNFCallStaticVoidMethod(env, jm_Config_refresh);
-    });
+    //    JNFPerformEnvBlock(JNFThreadDetachOnThreadDeath | JNFThreadSetSystemClassLoaderOnAttach | JNFThreadAttachAsDaemon, ^(JNIEnv *env) {
+    JNIEnv *env;
+    jint status = (*localVM)->GetEnv(localVM, (void**)&env, JNI_VERSION_1_2);
+    if (status == JNI_EDETACHED) {
+        status = (*localVM)->AttachCurrentThreadAsDaemon(localVM, (void**)&env, NULL);
+    }
+    if (status == 0) {
+        jclass jc_Config = (*env)->FindClass(env, "sun/security/krb5/Config");
+        CHECK_NULL(jc_Config);
+        jmethodID jm_Config_refresh = (*env)->GetStaticMethodID(env, jc_Config, "refresh", "()V");
+        CHECK_NULL(jm_Config_refresh);
+        (*env)->CallStaticVoidMethod(env, jc_Config, jm_Config_refresh);
+        if ((*env)->ExceptionOccurred(env) != NULL) {
+            (*env)->ExceptionClear(env);
+        }
+    }
+    (*localVM)->DetachCurrentThread(localVM);
```

Comment:
I think you only want to detach if you actually attached ! you don't want to be detaching VM threads.

Reply:
So I should remember how env was retrieved and only detach when it's from `AttachCurrentThreadAsDaemon`?

Reply:
Yes that is the idea. BTW which thread was it attached to in what you saw ?

Reply:
It's a new thread in the main group. After the method is called, it exits (`Thread::exit`).